File: sdk/typescript/src/wire.ts
import { Builder, ByteBuffer } from "flatbuffers";

export type WireFormat = "json" | "flatbuffers";

export enum MessageType {
  HELLO = 0,
  WELCOME = 1,
  PUBLISH = 2,
  SUBSCRIBE = 3,
  UNSUBSCRIBE = 4,
  EVENT = 5,
  ACK = 6,
  ERROR = 7,
  NOISE = 8,
}

const TYPE_TO_NAME: Record<number, string> = {
  [MessageType.HELLO]: "hello",
  [MessageType.WELCOME]: "welcome",
  [MessageType.PUBLISH]: "publish",
  [MessageType.SUBSCRIBE]: "subscribe",
  [MessageType.UNSUBSCRIBE]: "unsubscribe",
  [MessageType.EVENT]: "event",
  [MessageType.ACK]: "ack",
  [MessageType.ERROR]: "error",
  [MessageType.NOISE]: "noise",
};

const NAME_TO_TYPE: Record<string, number> = Object.fromEntries(
  Object.entries(TYPE_TO_NAME).map(([key, value]) => [value, Number(key)])
);

export type DecodedMessage = {
  msgType: string;
  payload: Record<string, unknown>;
};

export function encodeMessage(payload: Record<string, unknown>, fmt: WireFormat): Uint8Array {
  if (fmt === "json") {
    return new TextEncoder().encode(JSON.stringify(payload));
  }
  return encodeFlatbuffers(payload);
}

export function decodeMessage(raw: Uint8Array | string, fmt: WireFormat): DecodedMessage {
  if (fmt === "json") {
    const text = typeof raw === "string" ? raw : new TextDecoder().decode(raw);
    return decodeJson(text);
  }
  const bytes = typeof raw === "string" ? new TextEncoder().encode(raw) : raw;
  return decodeFlatbuffers(bytes);
}

function decodeJson(text: string): DecodedMessage {
  const payload = JSON.parse(text) as Record<string, unknown>;
  const msgType = payload.type;
  if (typeof msgType !== "string") {
    throw new Error("message type missing");
  }
  return { msgType, payload };
}

function encodeFlatbuffers(payload: Record<string, unknown>): Uint8Array {
  const msgType = payload.type;
  if (typeof msgType !== "string") {
    throw new Error("message type missing");
  }
  const msgTypeId = NAME_TO_TYPE[msgType];
  if (msgTypeId === undefined) {
    throw new Error("unknown message type");
  }
  const body = new TextEncoder().encode(JSON.stringify(payload));
  const builder = new Builder(body.length + 64);
  const payloadVec = builder.createByteVector(body);
  builder.startObject(2);
  builder.addFieldInt8(0, msgTypeId, 0);
  builder.addFieldOffset(1, payloadVec, 0);
  const msg = builder.endObject();
  builder.finish(msg);
  return builder.asUint8Array();
}

function decodeFlatbuffers(raw: Uint8Array): DecodedMessage {
  const bb = new ByteBuffer(raw);
  const table = bb.readInt32(bb.position()) + bb.position();
  const msgTypeOffset = fieldOffset(bb, table, 4);
  if (!msgTypeOffset) {
    throw new Error("message type missing");
  }
  const msgTypeId = bb.readInt8(table + msgTypeOffset);
  const payloadOffset = fieldOffset(bb, table, 6);
  let payloadBytes = new Uint8Array();
  if (payloadOffset) {
    const vec = table + payloadOffset;
    const vecStart = vec + bb.readInt32(vec);
    const length = bb.readInt32(vecStart);
    payloadBytes = raw.slice(vecStart + 4, vecStart + 4 + length);
  }
  const msgType = TYPE_TO_NAME[msgTypeId];
  if (!msgType) {
    throw new Error("unknown message type");
  }
  const payloadRaw = payloadBytes.length ? payloadBytes : new TextEncoder().encode("{}");
  const payload = JSON.parse(new TextDecoder().decode(payloadRaw)) as Record<string, unknown>;
  return { msgType, payload };
}

function fieldOffset(bb: ByteBuffer, table: number, vtableOffset: number): number {
  const vtable = table - bb.readInt32(table);
  const vtableLength = bb.readInt16(vtable);
  if (vtableOffset < vtableLength) {
    return bb.readInt16(vtable + vtableOffset);
  }
  return 0;
}

File: sdk/python/aether/wire.py
"""Wire encoding/decoding for JSON and FlatBuffers frames."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

import flatbuffers
from flatbuffers import encode, number_types, packer, table

WireFormat = Literal["json", "flatbuffers"]


class MessageType:
    HELLO = 0
    WELCOME = 1
    PUBLISH = 2
    SUBSCRIBE = 3
    UNSUBSCRIBE = 4
    EVENT = 5
    ACK = 6
    ERROR = 7
    NOISE = 8


_TYPE_TO_NAME = {
    MessageType.HELLO: "hello",
    MessageType.WELCOME: "welcome",
    MessageType.PUBLISH: "publish",
    MessageType.SUBSCRIBE: "subscribe",
    MessageType.UNSUBSCRIBE: "unsubscribe",
    MessageType.EVENT: "event",
    MessageType.ACK: "ack",
    MessageType.ERROR: "error",
    MessageType.NOISE: "noise",
}

_NAME_TO_TYPE = {value: key for key, value in _TYPE_TO_NAME.items()}


@dataclass
class DecodedMessage:
    msg_type: str
    payload: dict[str, Any]


def encode_message(payload: dict[str, Any], *, fmt: WireFormat) -> bytes:
    if fmt == "json":
        return _encode_json(payload)
    return _encode_flatbuffers(payload)


def decode_message(raw: bytes | str, *, fmt: WireFormat) -> DecodedMessage:
    if fmt == "json":
        return _decode_json(raw)
    return _decode_flatbuffers(raw)


def _encode_json(payload: dict[str, Any]) -> bytes:
    import json

    return json.dumps(payload, separators=(",", ":")).encode("utf-8")


def _decode_json(raw: bytes | str) -> DecodedMessage:
    import json

    if isinstance(raw, bytes):
        raw = raw.decode("utf-8")
    message = json.loads(raw)
    if not isinstance(message, dict):
        raise ValueError("message must be object")
    msg_type = message.get("type")
    if not isinstance(msg_type, str):
        raise ValueError("message type missing")
    return DecodedMessage(msg_type=msg_type, payload=message)


def _encode_flatbuffers(payload: dict[str, Any]) -> bytes:
    msg_type = payload.get("type")
    if not isinstance(msg_type, str):
        raise ValueError("message type missing")
    msg_type_id = _NAME_TO_TYPE.get(msg_type)
    if msg_type_id is None:
        raise ValueError("unknown message type")
    body = _encode_json(payload)
    builder = flatbuffers.Builder(len(body) + 64)
    payload_vec = builder.CreateByteVector(body)
    builder.StartObject(2)
    builder.PrependUint8Slot(0, msg_type_id, 0)
    builder.PrependUOffsetTRelativeSlot(1, payload_vec, 0)
    msg = builder.EndObject()
    builder.Finish(msg)
    return bytes(builder.Output())


def _decode_flatbuffers(raw: bytes | str) -> DecodedMessage:
    if isinstance(raw, str):
        raw = raw.encode("utf-8")
    buf = raw
    uoffset = encode.Get(packer.uoffset, buf, 0)
    tab = table.Table(buf, uoffset)
    type_offset = tab.Offset(4)
    if type_offset == 0:
        raise ValueError("message type missing")
    msg_type_id = tab.Get(number_types.Uint8Flags, type_offset + tab.Pos)
    payload_offset = tab.Offset(6)
    payload: bytes
    if payload_offset == 0:
        payload = b"{}"
    else:
        start = tab.Vector(payload_offset)
        length = tab.VectorLen(payload_offset)
        payload = tab.Bytes[start : start + length]
    msg_type = _TYPE_TO_NAME.get(msg_type_id)
    if msg_type is None:
        raise ValueError("unknown message type")
    decoded = _decode_json(payload)
    return DecodedMessage(msg_type=msg_type, payload=decoded.payload)
